### DISCUSSION TITLE
Deadlocks

### DISCUSSION THREAD

**Discussion by tomasoplt** [mrpmorris/Fluxor#404] [General]
We are developing Web application which has around 50 states. Some states has about 200 000 records. And Fluxor is in general under heavy load. We are firing around 10 000 actions after application start to fill the stores. In the library is a bug (deadlocks) which appears especially after application start. You have two synchronization objects in classes  Store and Dispatcher and they could generate deadlock in some circumstances. After application start when store is not fully initalized and lot of actions is Dispatched lot of deadlocks appears sometimes. 

Example:
- you launch first session of web application ( it works )
- second session ( still works )
- after fifth and further sessions server becomes slow and application stuck

 We will make some ivestigations and I will try report some progress if we find solution. Bug it would be good, if you could look for this problem also in parallel. It

**Comment by mrpmorris** [comment]
If you can whittle your app down to as simple a repro project as possible so we can just click a button and wait for the lock that would be useful.

**Reply by tomasoplt** [reply]
Here is an example project in which deadlock arrise almost every time: https://github.com/rjirasek/fluxor_deadlock
As an author of example project written in source code: one of possible fixes is to use one shared resource for dispatcher and store lock statements. Currently, there are two resourcres used for lock statements and this is the reason why deadlock happens.

**Reply by mrpmorris** [reply]
This was very helpful, thanks!

I've got a fix in for it,  but not yet sure when I'll release. 

Will you try referencing from the source to see if it solves your problem?

https://github.com/mrpmorris/fluxor/tree/407/PeteM-Deadlocks

**Reply by mrpmorris** [reply]
Did you manage to find time to try this?

**Reply by tomasoplt** [reply]
We will try. But I think you should write unit test which fails with your previous version and pass with new upgraded version. It is a must!  To find this error was about 40 mandays.

**Reply by mrpmorris** [reply]
Did it solve your problem?

**Reply by tomasoplt** [reply]
We have currently our private modified branch. This week we will test your modified version and I will let you know..

**Reply by mrpmorris** [reply]
Great!

**Reply by tomasoplt** [reply]
We tested your fix and it's OK. No deadlocks 👍

**Reply by mrpmorris** [reply]
Excellent.  I'll see when I can patch that in to a release. 
Thanks for your help!

**Comment by mrpmorris** [comment]
I have

**Comment by mfmadsen** [comment]
Hi @mrpmorris, anything I can do to help integrate this fix in a new release? We are seeing the same issue, so are eager to get this fix applied 🙂

**Comment by mrpmorris** [comment]
@mfmadsen Sorry, I've been busy with work and this slipped my mind.
I'll see if I can get 5.8 released this weekend.

**Reply by mfmadsen** [reply]
No worries. And thank you so much for an awesome library!

**Comment by mrpmorris** [comment]
Done!
